paasta_tools/kubernetes_tools.py

```diff
@@ -4325,21 +4325,42 @@ def ensure_service_account(
     kube_client: KubeClient,
     k8s_role: Optional[str] = None,
 ) -> None:
+    role_annotation = "eks.amazonaws.com/role-arn"
     sa_name = get_service_account_name(iam_role, k8s_role)
 
-    if not any(
-        sa.metadata and sa.metadata.name == sa_name
-        for sa in get_all_service_accounts(kube_client, namespace)
-    ):
+    existing_sa = None
+    for sa in get_all_service_accounts(kube_client, namespace):
+        if sa.metadata and sa.metadata.name == sa_name:
+            existing_sa = sa
+            break
+    else:
         sa = V1ServiceAccount(
             kind="ServiceAccount",
             metadata=V1ObjectMeta(
                 name=sa_name,
                 namespace=namespace,
-                annotations={"eks.amazonaws.com/role-arn": iam_role},
+                annotations={role_annotation: iam_role},
             ),
         )
         kube_client.core.create_namespaced_service_account(namespace=namespace, body=sa)
+    if existing_sa:
+        requires_patch = False
+        if (
+            sa.metadata.annotations
+            and sa.metadata.annotations.get(role_annotation, None) != iam_role
+        ):
+            sa.metadata.annotations[role_annotation] = iam_role
+            requires_patch = True
+        elif sa.metadata.annotations is None:
+            sa.metadata.annotations = {role_annotation: iam_role}
+            requires_patch = True
```

Comment:
```suggestion
        if (
            not sa.metadata.annotations
            or sa.metadata.annotations.get(role_annotation, None) != iam_role
        ):
            # NOTE: we don't annotate SAs apart with anything other
            # than the pod identity role ARN, so this will remove
            # any annotations that folks may have manually added
            sa.metadata.annotations = {role_annotation: iam_role}
            requires_patch = True
```

should do more or less the same thing since we don't really expect any other annotations to be present :)

Reply:
Ack, thanks for confirming the extra logic isn't needed.